### PR TITLE
:sparkles: Refactor `generic-external-provider`

### PR DESCRIPTION
**Pull request by JonahSussman**
Refactored the `generic-external-provider`. PTAL the README

### CODE REVIEW

File: lsp/base_service_client/base_capabilities.go [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+
+	symbols := sc.GetAllDeclarations(ctx, sc.BaseConfig.WorkspaceFolders, query)
+
+	incidents := []provider.IncidentContext{}
+	incidentsMap := make(map[string]provider.IncidentContext) // Remove duplicates
+
+	for _, s := range symbols {
+		references := sc.GetAllReferences(ctx, s.Location.Value.(protocol.Location))
+
+	ITERATE_REFERENCES:
```
Comment (shawn-hurley) [inline lsp/base_service_client/base_capabilities.go:63]: remove the goto please :)

Comment (JonahSussman) [reply lsp/base_service_client/base_capabilities.go]: Give it a chance! Alas, the goto is merely misunderstood...

Removed it

File: event/bench_test.go [hunk 2]
```diff
@@ -0,0 +1,158 @@
+// Copyright 2020 The Go Authors. All rights reserved.
```
Comment (fabianvf) [inline event/bench_test.go:1]: can you provide some background on this event package?

Comment (JonahSussman) [reply event/bench_test.go:1]: [`jsonrpc2_v2` is planned by the gopls team to go public](https://github.com/golang/go/issues/46187#issuecomment-842445007), but one of the dependencies it has is this event package. Both are internal as of now, so I had to duplicate them both for it to work. Is there a more elegant solution?

File: examples/yaml/.custom.json.kate-swp
Comment (fabianvf) [inline examples/yaml/.custom.json.kate-swp:1]: is this an intentional file?

File: external-providers/generic-external-provider/pkg/server_configurations/generic/service_client_test.go [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+	if err != nil {
+		panic(err)
+	}
+
+	fmt.Printf("%s\n", string(res))
+
+	var calls []EvaluateCall
+	calls = append(
+		calls,
+		// EvaluateCall{pythonSC, "referenced", []byte(`{"referenced":{pattern: "def hello_world"}}`)},
```
Comment (fabianvf) [inline external-providers/generic-external-provider/pkg/server_configurations/generic/service_client_test.go:72]: should these be removed?

File: external-providers/generic-external-provider/pkg/server_configurations/pylsp/service_client.go [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+		return nil, err
+	}
+	sc.LSPServiceClientEvaluator = eval
+
+	return sc, nil
+}
+
+// Tidy aliases
+
+type scFn = base.LSPServiceClientFunc[*PythonServiceClient]
```
Comment (fabianvf) [inline external-providers/generic-external-provider/pkg/server_configurations/pylsp/service_client.go:79]: an alias isn't bad but let's make it a little more descriptive to make it easier to read when we encounter that alias

File: external-providers/generic-external-provider/pkg/server_configurations/pylsp/service_client.go [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+
+func scTc(v any) openapi3.SchemaRef {
```
Comment (fabianvf) [inline external-providers/generic-external-provider/pkg/server_configurations/pylsp/service_client.go:81]: Same here let's use longer more descriptive names

File: external-providers/generic-external-provider/pkg/server_configurations/pylsp/service_client_test.go [hunk 6]
```diff
@@ -0,0 +1,211 @@
...
+	}
+
+	diagnostics := pylspSC.(*generic.GenericServiceClient).
+		PublishDiagnosticsCache.
+		Get("/home/jonah/Projects/analyzer-lsp/examples/python/file_b.py").
+		Await()
+
+	fmt.Printf("Diagnostics: %v\n", diagnostics)
+
+	// time.Sleep(5 * time.Second)
```
Comment (fabianvf) [inline external-providers/generic-external-provider/pkg/server_configurations/pylsp/service_client_test.go:85]: should this be removed?

File: external-providers/generic-external-provider/pkg/server_configurations/yaml_language_server/service_client.go [hunk 7]
```diff
@@ -0,0 +1,324 @@
...
+	if err != nil {
+		return nil, err
+	}
+	sc.LSPServiceClientEvaluator = eval
+
+	return sc, nil
+}
+
+// Tidy alias
+type scFn = base.LSPServiceClientFunc[*YamlServiceClient]
```
Comment (fabianvf) [inline external-providers/generic-external-provider/pkg/server_configurations/yaml_language_server/service_client.go:114]: same comments here

**Review by fabianvf — CHANGES_REQUESTED** [review]

File: external-providers/generic-external-provider/pkg/server_configurations/generic/service_client_test.go (hunk 3 above)
Comment (JonahSussman) [reply external-providers/generic-external-provider/pkg/server_configurations/generic/service_client_test.go:72]: For now I'm good with removing them. At a later date we'll probably need some proper tests written up

File: external-providers/generic-external-provider/pkg/server_configurations/nodejs/service_client.go [hunk 8]
```diff
@@ -130,14 +130,14 @@ func (sc *NodeServiceClient) EvaluateReferenced(ctx context.Context, cap string,
 		return resp{}, fmt.Errorf("unable to get query info")
 	}
 
-	// f, err := os.ReadFile("/home/jonah/Projects/analyzer-lsp/examples/nodejs/test.ts")
+	// f, err := os.ReadFile("/path/to/test.ts")
```
[the remark this answers is not in the document]
Comment (JonahSussman) [reply external-providers/generic-external-provider/pkg/server_configurations/nodejs/service_client.go:133]: This is a really bad habit I gotta get over :weary:

File: external-providers/generic-external-provider/pkg/server_configurations/generic/service_client_test.go [hunk 9]
```diff
@@ -1,169 +0,0 @@
...
-	if err != nil {
-		panic(err)
-	}
-
-	fmt.Printf("%s\n", string(res))
-
-	var calls []EvaluateCall
-	calls = append(
-		calls,
-		// EvaluateCall{pythonSC, "referenced", []byte(`{"referenced":{pattern: "def hello_world"}}`)},
```
[the remark this answers is not in the document]
Comment (JonahSussman) [reply external-providers/generic-external-provider/pkg/server_configurations/generic/service_client_test.go]: Whoops, totally misinterpreted!

**Comment by pranavgaikwad** [comment]
I am closing this in favor of #512

**Pull request closed without merging**